lib/osbuild.py

```diff
@@ -0,0 +1,370 @@
+            if vals:
+                strings = ','.join(f'"{x}"' for x in vals)
+                self.assembled += f'{name} = [ {strings} ]\n'
+
+    @contextlib.contextmanager
+    def to_tmpfile(self):
+        bp = self.assembled
+        util.log(f"using blueprint:\n{textwrap.indent(bp, '    ')}")
+        with tempfile.NamedTemporaryFile(mode='w') as f:
+            f.write(bp)
+            f.flush()
+            yield Path(f.name)
+
+    @contextlib.contextmanager
+    def to_composer(self):
+        blueprints = composer_cli_out('blueprints', 'list', log=False)
+        if self.NAME in blueprints:
+            composer_cli('blueprints', 'delete', self.NAME)
+        with self.to_tmpfile() as f:
+            composer_cli('blueprints', 'push', f)
+        yield self.NAME
+        composer_cli('blueprints', 'delete', self.NAME)
+
+
+class Guest(virt.Guest):
+    """
+    TODO document
+    """
+    DATASTREAM = Path(util.RPMPACK_DATA) / 'ds.xml'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.osbuild_log = f'{virt.GUEST_IMG_DIR}/{self.name}-osbuild.txt'
+
+    def wipe(self):
+        super().wipe()
+        log = Path(self.osbuild_log)
+        if log.exists():
+            os.remove(log)
+
+    def install(*args, **kwargs):
+        raise NotImplementedError("install() is not supported, use create()")
+
+    def create(self, *, blueprint=None, bp_verbatim=None, profile=None):
+        """
+        ... creates an image .. imports via virt-install
+        """
+        util.log(f"creating guest {self.name}")
+
+        # remove any previously installed guest
+        self.wipe()
+
+        if not blueprint:
+            blueprint = Blueprint()
+
+        if not bp_verbatim:
+            blueprint.add_host_repositories()
+            # copy our default package list from virt.Kickstart
+            for pkg in virt.Kickstart.PACKAGES:
+                blueprint.add_package(pkg)
+            # copy partitions too
+            for mountpoint, size in virt.Kickstart.PARTITIONS:
+                blueprint.add_partition(mountpoint, size*1024*1024)
+            # generate the ssh key on the same place as virt.Guest
+            util.ssh_keygen(self.ssh_keyfile_path)
+            with open(f'{self.ssh_keyfile_path}.pub') as f:
+                pubkey = f.read().rstrip()
+            blueprint.add_user('root', password=virt.GUEST_LOGIN_PASS, ssh_pubkey=pubkey)
+            # add openscap hardening, honor global excludes
+            if profile:
+                blueprint.add_openscap(self.DATASTREAM, profile)
+                #excludes = conf.remediation_excludes.host_os  # TEMP, TODO
+                excludes = []
+                blueprint.add_openscap_tailoring(unselected=excludes)
+
+        http_port = 8091
+        disk_path = Path(f'{virt.GUEST_IMG_DIR}/{self.name}.img')
+
+        with contextlib.ExitStack() as stack:
```

Comment:
Similarly as for `lib/virt.py`, please document it somewhere (rpm package with the setup hack, repo, package in blueprint) . Also comments as you wrote in https://github.com/RHSecurityCompliance/contest/pull/91#discussion_r1463714137 (`# this is RPM part`, `# this is HTTP server`) would be nice.

Reply:
Sure, will do.